lib/auth/grpcserver_test.go

```diff
@@ -2532,3 +2534,103 @@ func TestExport(t *testing.T) {
 		})
 	}
 }
+
+// TestSAMLValidation tests that SAML validation does not perform an HTTP
+// request if the calling user does not have permissions to create or update
+// a SAML connector.
+func TestSAMLValidation(t *testing.T) {
+	modules.SetTestModules(t, &modules.TestModules{
+		TestFeatures: modules.Features{SAML: true},
+	})
+
+	// minimal entity_descriptor to pass validation. not actually valid
+	const minimalEntityDescriptor = `
+<md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" entityID="http://example.com">
+  <md:IDPSSODescriptor>
+    <md:SingleSignOnService Location="http://example.com" />
+  </md:IDPSSODescriptor>
+</md:EntityDescriptor>`
+
+	allowSAMLUpsert := types.RoleConditions{
+		Rules: []types.Rule{{
+			Resources: []string{types.KindSAML},
+			Verbs:     []string{types.VerbCreate, types.VerbUpdate},
+		}},
+	}
+
+	testCases := []struct {
+		desc             string
+		allow            types.RoleConditions
+		entityDescriptor string
+		edURLCalled      bool
+		errAs            error
+	}{
+		{
+			desc:        "access denied",
+			allow:       types.RoleConditions{},
+			edURLCalled: false,
+			errAs:       &trace.AccessDeniedError{},
+		},
+		{
+			desc:             "validation failure",
+			allow:            allowSAMLUpsert,
+			entityDescriptor: "<unparsable XML",
+			edURLCalled:      true,
+			errAs:            &trace.BadParameterError{},
+		},
+		{
+			desc:             "access permitted",
+			allow:            allowSAMLUpsert,
+			entityDescriptor: minimalEntityDescriptor,
+			edURLCalled:      true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+
+			srv := newTestTLSServer(t)
```

Comment:
Maybe:

```suggestion
			server := newTestTLSServer(t)
```

(or `tlsServer`)

Reply:
done